js/modules/k6/grpc/client.go

```diff
@@ -212,15 +206,73 @@ func (t transportCreds) ClientHandshake(ctx context.Context,
 	return out, auth, err
 }
 
+func (c *Client) dial(
+	ctx context.Context,
+	addr string,
+	reflect bool,
+	state *lib.State,
+	isPlaintext bool,
+	errc chan error,
+) {
+	opts := []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.FailOnNonTempDialError(true),
+		grpc.WithStatsHandler(c),
+	}
+
+	if ua := state.Options.UserAgent; ua.Valid {
+		opts = append(opts, grpc.WithUserAgent(ua.ValueOrZero()))
+	}
+
+	if !isPlaintext {
+		tlsCfg := state.TLSConfig.Clone()
+		tlsCfg.NextProtos = []string{"h2"}
+
+		// TODO(rogchap): Would be good to add support for custom RootCAs (self signed)
+
+		// (rogchap) we create a wrapper for transport credentials so that we can report
+		// on any TLS errors.
+		creds := transportCreds{
+			credentials.NewTLS(tlsCfg),
+			errc,
+		}
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	}
+
+	if isPlaintext {
+		opts = append(opts, grpc.WithInsecure())
+	}
+
+	dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+		return state.Dialer.DialContext(ctx, "tcp", addr)
+	}
+	opts = append(opts, grpc.WithContextDialer(dialer))
+
+	var err error
+	c.conn, err = grpc.DialContext(ctx, addr, opts...)
+	if err != nil {
+		errc <- err
+		return
+	}
+	if reflect {
+		err := c.reflect(ctx)
+		if err != nil {
+			errc <- err
+			return
+		}
+	}
+
+	errc <- nil
+}
+
 // Connect is a block dial to the gRPC server at the given address (host:port)
-// nolint:funlen,cyclop
-func (c *Client) Connect(ctxPtr *context.Context, addr string, params map[string]interface{}) (bool, error) {
-	state := lib.GetState(*ctxPtr)
+func (c *Client) Connect(addr string, params map[string]interface{}) (bool, error) {
+	state := c.vu.State()
```

Comment:
I noticed that there are several places (here and other parts of there code) where we use this check:
```go
state := c.vu.State()
if state == nil {
	return false, errConnectInInitContext
}
```

How about adding a simple method like `modules.IsInitialContext(vu VU) bool` I believe that improves the readability of the code, and what is also essential make code more maintainable since we consolidate the logic that identifies that the particular place is dependent during what phase it is executing?

Reply:
I like the idea in terms of readability, but in this way, we can't reuse the `state` variable so for most of the methods we are doing 2x on the call for resolving the `State` that comes from the slow context https://github.com/grafana/k6/blob/0dda328d9a7f563607dae4fe89ff8c5138f8dc09/js/initcontext.go#L169-L171.

so I'm feeling a bit conflicted here without having some profiling data

Reply:
@olegbespalov please make a separate issue about this. 
@codebien in #2228 I in practice stop getting the state from a context, which is a thing that needs to happen for *everything* at some point either way.

But I do think this should be made ina *separate* PR and probably *across* the codebase, so please open an issue @olegbespalov so we can discuss the API for example and whether we want to do this for more stuff

Reply:
Just to keep the connection, opened an issue https://github.com/grafana/k6/issues/2368